src/main/java/io/cdap/plugin/SFTPCopyAction.java

```diff
@@ -151,53 +152,85 @@ public void run(ActionContext context) throws Exception {
     } catch(Exception e) {
-      throw new RuntimeException(String.format("Error occurred while connecting to SFTP Server %s", e.getMessage(), e));
+      throw new RuntimeException(String.format("Error occurred while copying files: %s", e.getMessage()), e);
     } finally {
         if (sftpConnector != null) {
           sftpConnector.close();
         }
     }
   }
 
-  private void copySFTPFiles(FileSystem fileSystem, Path destination, SFTPConnector SFTPConnector,
+  /**
+   * copySFTPFiles recursively copies all files and subdirectories in a given directory.
+   *
+   * @param fileSystem The destination file system to copy to
+   * @param destination The destination base path to copy to
+   * @param connector The SFTP connector to use
+   * @param context The context for the action
+   * @throws SftpException If any SFTP errors occur
+   */
+  private void copySFTPFiles(FileSystem fileSystem, Path destination, SFTPConnector connector,
                             ActionContext context) throws SftpException, IOException {
-    ChannelSftp channelSftp = SFTPConnector.getSftpChannel();
-    Vector files = channelSftp.ls(config.getSrcDirectory());
-    List<String> filesCopied = new ArrayList<>();
-    for (int index = 0; index < files.size(); index++) {
-      Object obj = files.elementAt(index);
+    ChannelSftp channel = connector.getSftpChannel();
+    ArrayList<String> filesCopied = new ArrayList<>();
+    Vector entries = channel.ls(config.getSrcDirectory());
+    for (Object obj : entries) {
       if (!(obj instanceof ChannelSftp.LsEntry)) {
         continue;
       }
       ChannelSftp.LsEntry entry = (ChannelSftp.LsEntry) obj;
-      if (".".equals(entry.getFilename()) || "..".equals(entry.getFilename())) {
-        // ignore "." and ".." files
-        continue;
+      recursiveSFTPCopy(channel, fileSystem, Paths.get(config.getSrcDirectory()), destination, filesCopied, entry);
+    }
+    context.getArguments().set(config.getVariableNameHoldingFileList(), Joiner.on(",").join(filesCopied));
+    LOG.info("Variables copied to {}.", Joiner.on(",").join(filesCopied));
+  }
+
+  private String getLongPath(java.nio.file.Path prefixPath, String fileName) {
+    return prefixPath.resolve(fileName).toString();
+  }
+
+  private void recursiveSFTPCopy(ChannelSftp channel, FileSystem fileSystem, java.nio.file.Path prefixPath,
+                                 Path destinationPath, List<String> filesCopied, ChannelSftp.LsEntry entry)
+    throws SftpException, IOException {
+    if (".".equals(entry.getFilename()) || "..".equals(entry.getFilename())) {
+      // ignore "." and ".." files
+      return;
+    }
+    // Ignore files that don't match the given file regex
+    if (!Strings.isNullOrEmpty(config.fileNameRegex)) {
+      if (!entry.getFilename().matches(config.fileNameRegex)) {
+        LOG.debug("Skipping file {} since it doesn't match the regex.", getLongPath(prefixPath, entry.getFilename()));
```

Comment:
nit: try to avoid contractions in user facing messaging. doesn't -> does not

Reply:
Fixed